### DISCUSSION TITLE
Directly interoperate with Android code

### DISCUSSION THREAD

**Discussion by Sitin** [fzyzcjy/flutter_rust_bridge#2464] [General]
What is the most effective way to interoperate with Android platform-specific code without Flutter?

Is there any bus, stream, IPC, or event queue I can use?

## Context

I am working on an app that works with serial ports via USB. Unfortunately I haven't found a nice way to connect to serial ports on Android. Meanwhile, I know how to work with serial over USB using Java/Kotlin. I want all my logic reside in Rust and keep only a thin Android handler that handles USB devices, basic 
I/O and passes everything to my Rust application.

**Comment by fzyzcjy** [comment]
Hi, maybe something like https://stackoverflow.com/questions/56821728/how-can-i-invoke-a-java-method-from-rust-via-jni, e.g. the j4rs package it mentions, may help.
